### PR TITLE
V6 model with ability to cancel approved vacation

### PR DESCRIPTION
**Pull request by rohanjr**
No UI or upgrade path to this yet, just want to validate the model. Might be easier to review using individual commits.

If we agree on the design, is there a reason we can't merge this before the UI and upgrades?

### CODE REVIEW

File: v6/daml/DAVL/V6.daml [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+template Vacation with
+    employeeRole: EmployeeRole
+    fromDate: Date
+    toDate: Date
+    pendingCancel: Bool
+  where
+    let EmployeeRole{..} = employeeRole
+    signatory [employee, company, boss]
+
```
Comment (rohanjr) [inline v6/daml/DAVL/V6.daml]: Change from V5 is all here

File: v6/daml/DAVL/V6.daml [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    nonconsuming choice EmployeeRole_RequestVacation: ContractId VacationRequest with
+        fromDate: Date
+        toDate: Date
+      controller employee
+      do
+        let vacation = Vacation with
+              employeeRole = this
+              fromDate
+              toDate
+              pendingCancel = False
```
Comment (rohanjr) [inline v6/daml/DAVL/V6.daml]: Added field to the Vacation contract

File: v6/daml/DAVL/V6.daml [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        archive self
+        -- Add vacation days back to employee allocation
+        (allocationId, allocation) <- fetchByKey @EmployeeVacationAllocation (key employeeRole)
+        assertMsg "mismatching employee in vacation" $ employeeRole == allocation.employeeRole
+        let days = Date.subDate toDate fromDate + 1
+        let remainingDays = allocation.remainingDays + days
+        archive allocationId
+        create allocation with remainingDays
+
+createEmployee: Party -> Text -> Party -> Int -> Scenario Party
```
Comment (shayne-fletcher) [inline v6/daml/DAVL/V6.daml]: Perhaps these utilities should move to a `where` clause under the scenario?

File: v6/daml/DAVL/V6.daml (hunk 2 above)
Comment (hurryabit) [reply v6/daml/DAVL/V6.daml]: As I said elsewhere, I'm totally not a fan of this for reasons I gave there as well.

File: v6/daml/DAVL/V6.daml [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+  vacation <- submit shaul do
+    fetch vacationId
+  (_, shaulEmployee) <- submit shaul do
+    fetchByKey @EmployeeRole shaul
+  assert $ vacation.employeeRole == shaulEmployee
+
+  assertRemainingDays shaul 20
+
+  -- Cancellation requests
```
Comment (rohanjr) [inline v6/daml/DAVL/V6.daml]: Also added to the scenario here

File: v6/daml/DAVL/V6.daml [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        validateVacation = do
+          vacation' <- fetch vacationId
+          assertMsg "Vacation ID does not match provided vacation details" $ vacation' == vacation
+
+    signatory [employee, company]
+    observer boss
+
+    -- Ensure at most one cancel request per vacation.
+    key vacation: Vacation
+    maintainer key.employeeRole.employee
```
Comment (hurryabit) [inline v6/daml/DAVL/V6.daml]: We don't have a corresponding key on `Vacation`, thus it feels strange to have key here. It's debatable whether we should have a key on `Vacation` but let's keep that for a later feature addition.
```suggestion
```

File: v6/daml/DAVL/V6.daml [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+  assert $ vacation.employeeRole == shaulEmployee
+
+  assertRemainingDays shaul 20
+
+  -- Cancellation requests
+  cancelRequest <- submit shaul do
+    exercise vacationId Vacation_RequestCancel
+  -- Cannot request to cancel same vacation twice
+  submitMustFail shaul do
+    exercise vacationId Vacation_RequestCancel
```
Comment (hurryabit) [inline v6/daml/DAVL/V6.daml]: As I said above, let's not do this for now.
```suggestion
```

**Pull request closed without merging**